src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1610 @@
+                    }
+                }
+                AstElem::PkH(ref pkh) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    let res = self.stack.evaluate_pkh(self.secp, self.sighash, pkh);
+                    if res.is_some(){
+                        return res;
+                    }
+                }
+                AstElem::After(ref n) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    let res = self.stack.evaluate_after(n, self.age);
+                    if res.is_some(){
+                        return res;
+                    }
+                }
+                AstElem::Older(ref n) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    let res = self.stack.evaluate_older(n, self.height);
+                    if res.is_some(){
+                        return res;
+                    }
+                }
+                AstElem::Sha256(ref hash) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    let res = self.stack.evaluate_sha256(hash);
+                    if res.is_some(){
+                        return res;
+                    }
+                }
+                AstElem::Hash256(ref hash) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    let res = self.stack.evaluate_hash256(hash);
+                    if res.is_some() {
+                        return res;
+                    }
+                }
+                AstElem::Hash160(ref hash) =>{
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    let res = self.stack.evaluate_hash160(hash);
+                    if res.is_some(){
+                        return res;
+                    }
+                }
+                AstElem::Ripemd160(ref hash) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    let res = self.stack.evaluate_ripemd160(hash);
+                    if res.is_some(){
+                        return res;
+                    }
+                }
+                AstElem::Alt(ref sub) |
+                AstElem::Swap(ref sub) |
+                AstElem::Check(ref sub) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    self.state.push(NodeEvaluationState { node: sub, n_evaluated: 0, n_satisfied: 0 });
```

Comment:
I'd implement a helper function `self.push_evaluation_state(sub, 0, 0)` to save yourself a bunch of typing. But I don't have a strong preference